common/lib/dependabot/new_version.rb

```diff
@@ -0,0 +1,228 @@
+      version = version.gsub(/(\d)([A-Za-z])/, '\1-\2')
+      version = version.gsub(/([A-Za-z])(\d)/, '\1-\2')
+
+      # Replace empty tokens with 0
+      version = version.gsub(/([\.\-])([\.\-])/, '\10\2')
+      version = version.gsub(/^([\.\-])/, '0\1')
+      version.gsub(/([\.\-])$/, '\10')
+    end
+
+    sig { params(version: String).returns(String) }
+    def trim_version(version)
+      version.split("-").filter_map do |v|
+        parts = T.let(v.split("."), T.nilable(T::Array[String]))
+        parts = T.must(parts)[0..-2] while NULL_VALUES.include?(T.must(parts).last)
+        T.must(parts).join(".")
+      end.reject(&:empty?).join("-")
+    end
+
+    sig { params(version: String, other_version: String).returns(T::Array[String]) }
+    def convert_dates(version, other_version)
+      default = [version, other_version]
+      return default unless version.match?(/^\d{4}-?\d{2}-?\d{2}$/)
+      return default unless other_version.match?(/^\d{4}-?\d{2}-?\d{2}$/)
+
+      [version.delete("-"), other_version.delete("-")]
+    end
+
+    sig { params(version: String).returns(T::Array[String]) }
+    def split_into_prefixed_tokens(version)
+      ".#{version}".split(/(?=[\-\.\+])/)
+    end
+
+    sig do
+      params(prefixed_tokens: T::Array[String],
+             other_prefixed_tokens: T::Array[String]).returns(T::Array[T::Array[String]])
+    end
+    def pad_for_comparison(prefixed_tokens, other_prefixed_tokens)
+      prefixed_tokens = prefixed_tokens.dup
+      other_prefixed_tokens = other_prefixed_tokens.dup
+
+      longest = [prefixed_tokens, other_prefixed_tokens].max_by(&:count)
+      shortest = [prefixed_tokens, other_prefixed_tokens].min_by(&:count)
+
+      T.must(longest).count.times do |index|
+        next unless T.must(shortest)[index].nil?
+
+        segment = T.must(T.must(longest)[index])
+
+        T.must(shortest)[index] = segment.start_with?(".") ? ".0" : "-"
+      end
+
+      [prefixed_tokens, other_prefixed_tokens]
+    end
+
+    sig { params(prefix: String, token: String, other_prefix: String, other_token: String).returns(T.nilable(Integer)) }
+    def compare_prefixed_token(prefix:, token:, other_prefix:, other_token:)
+      token_type = token.match?(/^\d+$/) ? :number : :qualifier
+      other_token_type = other_token.match?(/^\d+$/) ? :number : :qualifier
+
+      hierarchy = PREFIXED_TOKEN_HIERARCHY.fetch(prefix).fetch(token_type)
+      other_hierarchy =
+        PREFIXED_TOKEN_HIERARCHY.fetch(other_prefix).fetch(other_token_type)
+
+      hierarchy_comparison = hierarchy <=> other_hierarchy
+      return hierarchy_comparison unless hierarchy_comparison.zero?
+
+      compare_token(token: token, other_token: other_token)
+    end
+
+    sig { params(token: String, other_token: String).returns(T.nilable(Integer)) }
+    def compare_token(token:, other_token:)
+      if (token_hierarchy = NAMED_QUALIFIERS_HIERARCHY[token])
+        return -1 unless NAMED_QUALIFIERS_HIERARCHY[other_token]
```

Comment:
can we have unit tests covering all these?

Reply:
For the public interface methods, I agree it would probably be better to add tests. However, for now I'd just duplicate the tests already in Maven and that's why I didn't add them. I prefer to wait until I update a few more ecosystems and get a good feel of what is generic enough to test in the superclass and what is ecosystem specific.

Reply:
can we move the tests in maven over as part of this migration too? That way we avoid regressions and I'd like to have tests included so we do not introduce regressions as part of this effort

Reply:
The main public methods are being covered in the test file I copied from maven.
<img width="353" alt="Screenshot 2024-08-07 at 17 08 58" src="https://github.com/user-attachments/assets/40727069-7e37-44ee-83a8-38288cedd942">